src/Microsoft.AspNet.Mvc.Core/Formatters/MediaTypeEncoding.cs

```diff
@@ -0,0 +1,91 @@
+        /// Gets the <see cref="Encoding"/> for the given <see cref="mediaType"/> if it exists.
+        /// </summary>
+        /// <param name="mediaType">The mediaType from which to get the charset parameter.</param>
+        /// <returns>The <see cref="Encoding"/> of the mediaType if it exists; otherwise <code>null</code>.</returns>
+        public static Encoding GetEncoding(StringSegment mediaType)
+        {
+            var charset = GetCharset(mediaType);
+            return GetEncodingFromCharset(charset);
+        }
+
+        /// <summary>
+        /// Gets the <see cref="Encoding"/> for the given <see cref="mediaType"/> if it exists.
+        /// </summary>
+        /// <param name="mediaType">The mediaType from which to get the charset parameter.</param>
+        /// <returns>The <see cref="Encoding"/> of the mediaType if it exists or a <see cref="StringSegment"/> without value if not.</returns>
+        public static Encoding GetEncoding(string mediaType)
+        {
+            var charset = GetCharset(new StringSegment(mediaType));
+            return GetEncodingFromCharset(charset);
+        }
+
+        /// <summary>
+        /// Gets the charset parameter of the given <paramref name="mediaType"/> if it exists.
+        /// </summary>
+        /// <param name="mediaType">The mediaType from which to get the charset parameter.</param>
+        /// <returns>The charset of the mediaType if it exists or a <see cref="StringSegment"/> without value if not.</returns>
+        public static StringSegment GetCharset(StringSegment mediaType)
+        {
+            try
+            {
+                var mt = MediaTypeHeaderValue.Parse(mediaType.Value);
+                return new StringSegment(mt.Charset);
+
+            }
+            catch (Exception)
+            {
+
+                return new StringSegment();
+            }
+        }
+
+        public static string ReplaceEncoding(string mediaType, Encoding encoding)
+        {
+            return ReplaceEncoding(new StringSegment(mediaType), encoding);
+        }
+
+        public static string ReplaceEncoding(StringSegment mediaType, Encoding encoding)
+        {
+            var charset = GetCharset(mediaType);
+
+            if (GetEncodingFromCharset(charset) == encoding)
+            {
+                return mediaType.Value;
+            }
+
+            var mt = MediaTypeHeaderValue.Parse(mediaType.Value);
+            mt.Encoding = encoding;
+
+            return mt.ToString();
+        }
+
+        private static Encoding GetEncodingFromCharset(StringSegment charset)
+        {
+            if (charset.Equals("utf-8", StringComparison.OrdinalIgnoreCase))
+            {
+                // This is an optimization for utf-8 that prevents the Substring caused by
+                // charset.Value
+                return Encoding.UTF8;
+            }
+
+            try
+            {
+                return charset.HasValue ? Encoding.GetEncoding(charset.Value) : null;
```

Comment:
There's no way to `TryGetEncoding` in the BCL, which is unfortunate. If our only alternative is to call something that might throw, we should leave a comment.